android/src/main/kotlin/com/github/fluttercommunity/flutterdownloader/KotlinDownload.kt

```diff
@@ -0,0 +1,97 @@
+package com.github.fluttercommunity.flutterdownloader
+
+import androidx.annotation.UiThread
+import io.flutter.plugin.common.MethodChannel
+import java.io.File
+import java.net.URL
+
+/**
+ * The Kotlin part of the PlatformDownload from dart. The naming should be quiet similar to the dart and Swift
+ * implementation. Some details will differ since that are different programming languages.
+ */
+internal class KotlinDownload(
+    /// The urlHash used as internal id of the Download
+    val urlHash: String
+) {
+    /// The cache file of the (partial) download
+    val cacheFile = File(FlutterDownloaderPlugin.tempDir, "$urlHash.part")
+
+    /// The request headers
+    val headers = mutableMapOf<String, String>()
+    val url: URL
+
+    /// The filename which should be used for the filesystem
+    var filename: String? = null
+        private set
+
+    /// The etag if given to resume the download
+    var eTag: String? = null
+        private set
+    val target: Target
+
+    /// The file size of the file to download
+    var finalSize: Long? = null
+        @UiThread
+        set(value) {
+            field = value
+            backChannel.invokeMethod("updateSize", value)
+        }
+    var progress: Long = 0
+        @UiThread
+        set(value) {
+            field = value
+            backChannel.invokeMethod("updateProgress", value)
+        }
+    var status: DownloadStatus = DownloadStatus.paused
+        @UiThread
+        set(value) {
+            field = value
+            backChannel.invokeMethod("updateStatus", value.name)
+        }
+
+    private val backChannel = FlutterDownloaderPlugin.getBackChannel(urlHash)
+
+    init {
+        /// Parse meta file
```

Comment:
Could we use JSON for the meta file instead of a custom key=value pair?

Reply:
Likely a better choice, but I json convertion trends to be a bit complicated to setup on each platform.

Reply:
For Flutter side, we can use json_serializable and freezeed.

For Android, we can use kotlinx.serialization.

For iOS, we can use the Codable protocol.

I've used all three and they work nicely:)

Reply:
I put it to the TODO list

Reply:
Can you share a copy&paste sample for that Codable? Otherwise I will google that myself when I'm again on the iOS part :-)